tools/nl/embeddings/utils.py

```diff
@@ -13,146 +13,192 @@
-  return alts
-
-
-def add_sv(name: str, sv: str, text2sv: Dict[str, str],
-           dup_svs: List[List[str]]) -> None:
-  osv = text2sv.get(name)
-  if not osv or osv == sv:
-    text2sv[name] = sv
-    return
-
-  # This is a case of duplicate SV.  Prefer the non sdg, human-curated, shorter SV.
-  # Track it.
-  pref, drop = sv, osv
-  if sv.startswith('dc/topic/sdg'):
-    # sv is an sdg topic. Prefer osv if osv is not an sdg topic. Otherwise, go
-    # by dcid len.
-    if not osv.startswith('dc/topic/sdg') or len(osv) <= len(sv):
-      pref, drop = osv, sv
-  elif ((osv.startswith('dc/') and sv.startswith('dc/')) or
-        (not osv.startswith('dc/') and not sv.startswith('dc/'))):
-    # Both SVs are autogen or both aren't. Go by dcid len.
-    if len(osv) <= len(sv):
-      pref, drop = osv, sv
-  elif sv.startswith('dc/'):
-    # sv is autogen, prefer osv.
-    pref, drop = osv, sv
-
-  text2sv[name] = pref
-  dup_svs.append([pref, drop, name])
-
-
-def dedup_texts(df: pd.DataFrame) -> Tuple[Dict[str, str], List[List[str]]]:
-  """Dedup multiple texts mapped to the same DCID and return a list."""
-  text2sv_dict = {}
-  dup_sv_rows = [['PreferredSV', 'DroppedSV', 'DuplicateName']]
-  for _, row in df.iterrows():
-    sv = row[DCID_COL].strip()
-
-    # All alternative sentences are retrieved from COL_ALTERNATIVES, which
-    # are expected to be delimited by ";" (semi-colon).
-    if COL_ALTERNATIVES in row:
-      alternatives = row[COL_ALTERNATIVES].split(';')
-      alternatives = [a.strip() for a in alternatives if a.strip()]
-      for alt in alternatives:
-        add_sv(alt, sv, text2sv_dict, dup_sv_rows)
-
-  return (text2sv_dict, dup_sv_rows)
-
-
-def build_embeddings(
-    text2sv: Dict[str, str],
-    model: SentenceTransformer = None,
-    model_endpoint: aiplatform.Endpoint = None) -> pd.DataFrame:
-  """Builds the embeddings dataframe.
-
-  The output dataframe contains the embeddings columns (typically 384) + dcid + sentence.
+def build_preindex(fm: FileManager) -> Tuple[List[str], List[str]]:
+  """
+  Build preindex records (sentence -> dcid) from a directory of CSV files.
   """
-  texts = sorted(list(text2sv.keys()))
-
-  if model:
-    embeddings = model.encode(texts, show_progress_bar=True)
-  else:
-    embeddings = []
-    for i, chuck in enumerate(chunk_list(texts, _CHUNK_SIZE)):
-      logging.info('texts %d to %d', i * _CHUNK_SIZE, (i + 1) * _CHUNK_SIZE - 1)
-      for i in range(_MODEL_ENDPOINT_RETRIES):
-        try:
-          resp = model_endpoint.predict(instances=chuck,
-                                        timeout=600).predictions
-          embeddings.extend(resp)
-          break
-        except Exception as e:
-          logging.info('Exception %s', e)
-  embeddings = pd.DataFrame(embeddings)
-  embeddings[DCID_COL] = [text2sv[t] for t in texts]
-  embeddings[COL_ALTERNATIVES] = texts
+  text2sv: Dict[str, set[str]] = {}
+  for file_name in glob.glob(fm.local_input_dir() + "/[!_]*.csv"):
```

Comment:
This reads all CSVs from the input dir and not specific ones. Can we have a mode where it only looks for a specific CSV or CSVs that start with a certain prefix? For custom DCs we currently generate a `sentences.csv` in the nl folder and we only want to read that one. I was planning to generate other files (e.g. topics) in the same folder and we don't want them to get read here. (The topics files are not CSVs so we are ok for now but should future proof it if possible)

Reply:
Using a directory is more clear and with a few benefits:

1. It can easily support several csv files without changing any code or logic and avoid accidentally adding other non-related files with same prefix.
2. It's very easy for gcs <-> local two way download/upload
3. `setences.csv` can be treated as `sentences/embeddings.csv` directly.

Since this input file is purely internal and non-visible to users, I have updated it in admin/html.py and custom_catalog.yaml. We can sync Ajai's change when they both land?

Reply:
OK. We'll just need to be careful about generating any other CSVs in this folder for custom dcs. Currently it's a non-issue since `sentences.csv` is the only one.